Implement PREPARE

rschlussel opened the pull request:
This Pull request includes an implementation for PREPARE queries.  It is the first step for supporting prepared statements in presto.

Because Presto does not save session state, this was implemented in the same way that session variables are.  Prepared statements for a session are sent by the client via an http session header that the server parses and adds to its session.   When a prepare query is executed, it will be included in the response to the client via the X-Presto-Added-Prepare header so the client can maintain the session state.

Support for parameters is not added yet.

rschlussel commented:
@petroav

electrum commented:
Lots of minor issues, but overall seems simple and straight forward. This is a good start.

rschlussel commented:
I just noticed some small formatting issues still.  I'm going through it again and will update the commits addressing comments.

rschlussel commented:
updated

rschlussel commented:
Anything else?

rschlussel commented:
@electrum any other comments?  Let me know if it would be helpful for review to squash any of the commits together.

rschlussel commented:
@electrum Can you look at this again please?

rschlussel commented:
@martint @electrum can someone please review this again.

martint commented on presto-parser/src/main/java/com/facebook/presto/sql/tree/Prepare.java:
```diff
@@ -0,0 +1,84 @@
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+
+public class Prepare extends Statement
```
Please re-format according to https://github.com/airlift/codestyle

martint replied:
Oh, nevermind... I was reviewing an old commit

martint commented on presto-parser/src/main/java/com/facebook/presto/sql/tree/Prepare.java:
```diff
@@ -0,0 +1,84 @@
+            .toString();
+  }
+
+  public Query getQuery()
```
Move this method towards the top, close to getIdentifier.

martint commented on presto-main/src/main/java/com/facebook/presto/execution/CreateViewTask.java:
```diff
@@ -109,21 +108,20 @@ private Analysis analyzeStatement(Statement statement, Session session, Metadata
     }
 
-    private String getFormattedSql(CreateView statement)
+    public static String getFormattedSql(Statement statement, SqlParser sqlParser)
```
This method doesn't seem to belong here (especially if you're making it public so that it can be used elsewhere). Maybe move it to SqlFormatter.

rschlussel replied:
done

martint commented on presto-main/src/main/java/com/facebook/presto/server/StatementResource.java:
```diff
@@ -210,6 +216,23 @@ private static Response getQueryResults(Query query, Optional<Long> token, UriIn
+                        response.header(PRESTO_ADDED_PREPARE, encodedKey + '=' + encodedValue);
+                    }
+                    catch (UnsupportedEncodingException e) {
+                        throw new PrestoException(UNSUPPORTED_ENCODING, "Cannot encode statement: UTF-8 encoding unsupported");
```
This shouldn't be a PrestoException. If "UTF-8" is unsupported, something is really wrong with the underlying environment/VM. Throw an AssertionError instead.

rschlussel replied:
done

martint commented:
This PR has a bunch of changes related to "deallocate", but I don't see support for that in the grammar or execution. I'd defer those changes until that functionality is implemented, and make this exclusively about PREPARE.

rschlussel commented:
Updated.  I squashed the commits together so I could reasonably extract the functionality related to "deallocate" into the deallocate patch, which will follow this one.

rschlussel commented:
@electrum I've addressed all comments from you and @martint.  Please review.

electrum commented on presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java:
```diff
@@ -888,4 +903,23 @@ private static void appendAliasColumns(StringBuilder builder, List<String> colum
+            parsed = sqlParser.createStatement(sql);
+        }
+        catch (ParsingException e) {
+            throw new ParsingException("Formatted query does not parse: " + statement);
```
This throws the wrong exception. It would result in a confusing syntax error for the user when in reality it is an internal bug.

rschlussel replied:
PrestoException is not accessible from here without creating a dependency from presto-parser on presto-spi.  Is there a different exception I should use, or is it reasonable to create that dependency?

electrum replied:
The method shouldn't be moved to this module. Keep it in presto-main.

rschlussel replied:
k.  I moved it to a utility class in presto-main.

electrum commented on presto-main/src/main/java/com/facebook/presto/execution/PrepareTask.java:
```diff
@@ -0,0 +1,61 @@
+    @Override
+    public CompletableFuture<?> execute(Prepare prepare, TransactionManager transactionManager, Metadata metadata, AccessControl accessControl, QueryStateMachine stateMachine)
+    {
+        String statementName = prepare.getName();
```
This is only used once and is short, inline it

electrum commented on presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java:
```diff
@@ -393,6 +398,25 @@ public void addResetSessionProperties(String name)
+        requireNonNull(key, "key is null");
+        requireNonNull(value, "value is null");
+
+        if (session.getPreparedStatements().containsKey(key)) {
```
This check is not necessary, since the `putIfAbsent` will fail if it exists

rschlussel replied:
putIfAbsent goes to the addedPreparedStatements, which are the things to be updated as a result of this transaction.  But first you need to check that the session doesn't already have a prepared statement with that name, or you'll end up overwriting an existing statement, which is not allowed.

electrum replied:
Ah, I see

electrum commented on presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java:
```diff
@@ -393,6 +398,25 @@ public void addResetSessionProperties(String name)
+        if (session.getPreparedStatements().containsKey(key)) {
+            throw new PrestoException(ALREADY_EXISTS, "Prepared statement already exists: " + key);
+        }
+        String previousValue = addedPreparedStatements.putIfAbsent(key, value);
```
This variable can be inlined

electrum commented on presto-main/src/main/java/com/facebook/presto/server/ResourceUtil.java:
```diff
@@ -203,6 +214,33 @@ public static void assertRequest(boolean expression, String format, Object... ar
+            statementName = URLDecoder.decode(nameValue.get(0), "UTF-8");
+            sqlString = URLDecoder.decode(nameValue.get(1), "UTF-8");
+        }
+        catch (UnsupportedEncodingException e) {
```
`UnsupportedEncodingException` is only thrown if the encoding is not supported. This can be re-thrown as `AssertionError` since `UTF-8` is guaranteed to be supported.

We do need to catch `IllegalArgumentException` since that is thrown if the input data is invalid.

```java
catch (UnsupportedEncodingException e) {
    throw new AssertionError(e);
}
catch (IllegalArgumentException e) {
    throw badRequest(format("Invalid %s header: %s", PRESTO_PREPARED_STATEMENT, e.getMessage()));
}

electrum commented on presto-main/src/main/java/com/facebook/presto/server/ResourceUtil.java:
```diff
@@ -203,6 +214,33 @@ public static void assertRequest(boolean expression, String format, Object... ar
+            sqlParser.createStatement(sqlString);
+        }
+        catch (ParsingException e) {
+            throw badRequest(e.getMessage());
```
```java
throw badRequest(format("Invalid %s header: %s", PRESTO_PREPARED_STATEMENT, e.getMessage()));

electrum commented on presto-main/src/main/java/com/facebook/presto/server/StatementResource.java:
```diff
@@ -210,6 +213,19 @@ private static Response getQueryResults(Query query, Optional<Long> token, UriIn
+        query.getAddedPreparedStatements().entrySet().stream()
+                .forEach(entry -> {
+                    try {
+                        String encodedKey = encode(entry.getKey(), "UTF-8");
```
Don't static import `encode` here since the name is generic and it's not clear that this is doing URL encoding (unfortunately, the method was not named `urlEncode`)

electrum commented on presto-main/src/main/java/com/facebook/presto/server/StatementResource.java:
```diff
@@ -210,6 +213,19 @@ private static Response getQueryResults(Query query, Optional<Long> token, UriIn
 
+        // add added prepare statements
+        query.getAddedPreparedStatements().entrySet().stream()
+                .forEach(entry -> {
```
This stream isn't doing anything other than iterating, so write it as a normal for-each loop

electrum commented on presto-main/src/main/java/com/facebook/presto/server/StatementResource.java:
```diff
@@ -210,6 +213,19 @@ private static Response getQueryResults(Query query, Optional<Long> token, UriIn
+                        response.header(PRESTO_ADDED_PREPARE, encodedKey + '=' + encodedValue);
+                    }
+                    catch (UnsupportedEncodingException e) {
+                        throw new AssertionError("Cannot encode statement: UTF-8 encoding unsupported");
```
This can just be `new AssertionError(e)` since it won't ever happen

electrum commented on presto-main/src/test/java/com/facebook/presto/execution/TestPrepareTask.java:
```diff
@@ -0,0 +1,85 @@
+public class TestPrepareTask
+{
+    private final MetadataManager metadata = MetadataManager.createTestMetadataManager();
+    private final ExecutorService executor = newCachedThreadPool(daemonThreadsNamed("stage-executor-%s"));
```
This name is wrong. Should be something like "test-%s"

electrum commented on presto-main/src/test/java/com/facebook/presto/execution/TestPrepareTask.java:
```diff
@@ -0,0 +1,85 @@
+
+public class TestPrepareTask
+{
+    private final MetadataManager metadata = MetadataManager.createTestMetadataManager();
```
Static import

electrum commented on presto-main/src/test/java/com/facebook/presto/execution/TestPrepareTask.java:
```diff
@@ -0,0 +1,85 @@
+        QueryStateMachine stateMachine = QueryStateMachine.begin(new QueryId("query"), sqlString, session, URI.create("fake://uri"), false, transactionManager, executor);
+        Prepare prepare = new Prepare(statementName, query);
+        new PrepareTask(new SqlParser()).execute(prepare, transactionManager, metadata, new AllowAllAccessControl(), stateMachine);
+        return stateMachine;
```
Have this return the list of prepared statements

electrum commented on presto-main/src/test/java/com/facebook/presto/execution/TestPrepareTask.java:
```diff
@@ -0,0 +1,85 @@
+        assertEquals(stateMachine.getAddedPreparedStatements(), ImmutableMap.of(statementName, "SELECT *\nFROM\n  foo\n"));
+    }
+
+    @Test(expectedExceptions = PrestoException.class)
```
Do this with a try/catch and assert the error code

electrum commented on presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java:
```diff
@@ -1421,6 +1422,15 @@ public void testCall()
+    public void testPrepare()
+    {
+        assertStatement("PREPARE myquery FROM select * from foo",
+                                new Prepare("myquery", simpleQuery(
```
The indentation looks off here

electrum commented:
I'll fix up all these minor issues and merge it

electrum commented:
Merged, thanks!
